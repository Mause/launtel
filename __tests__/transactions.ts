import { createServer } from "vercel-node-server";
import listen from "test-listen";
import Axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { Server } from "net";
import moxios from "moxios";
import { sign } from "jsonwebtoken";

let server: Server;
let url: string;
let instance: AxiosInstance;

const SECRET = "SECRET";
process.env.JWT_SECRET = SECRET;

beforeAll(async () => {
  jest.mock("../support/axios");
  const axios = require("../support/axios");
  axios.default.create = (config?: AxiosRequestConfig) => {
    instance = Axios.create(config);
    moxios.install(instance);
    return instance;
  };

  const routeUnderTest = require("../api/transactions");
  server = createServer(routeUnderTest.default);
  url = await listen(server);
});

afterAll(() => {
  server.close();
  moxios.uninstall(instance);
});

it("should return the expected response", async () => {
  moxios.stubOnce("POST", /login/, {
    response: "",
    status: 307,
    statusText: "Redirect",
    headers: {
      location: "https://fake/redirected",
    },
  });
  moxios.stubOnce("POST", /redirected/, {
    status: 200,
    response: "",
    headers: {
      "set-cookie": "hello=world",
    },
  });
  moxios.stubOnce("GET", /.*/, {
    status: 200,
    statusText: "Ok",
    headers: {},
    response: `
    <div>
      <table>
        <thead>
          <th>Date</th>
          <th>Description</th>
          <th>Amount</th>
          <th/>
          <th>Balance</th>
        </thead>
        <tbody>
          <tr>
            day, month, year, time
            <td>1 January 2021 11:00</td>
            <td>Debit</td>
            <td>-1.99</td>
            <td/>
            <td>20.00</td>
          </tr>
        </tbody>
      </table>
    </div>
      `,
  });
  const response = await Axios.get(url, {
    headers: {
      Authorization: "Bearer " + sign({ aud: "authenticated" }, SECRET),
    },
  });
  expect(response.status).toBe(200);
  expect(response.data).toMatchSnapshot();
});
