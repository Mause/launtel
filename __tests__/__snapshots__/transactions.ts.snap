// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the expected response 1`] = `
Object {
  "numTransactions": 1,
  "perMonth": Object {
    "2021-01": Object {
      "amount": 1.99,
      "discounted": 0,
    },
  },
  "transactions": Array [
    Object {
      "amount": -1.99,
      "balance": 20,
      "date": "2021-01-01T11:00",
      "description": "Debit",
      "isCCCharge": false,
    },
  ],
}
`;
